### DISCUSSION TITLE
translate -bounds uses origin?

### DISCUSSION THREAD

**Discussion by rgkirch** [phronmophobic/membrane#43] [Q&A]
```clojure
(defrecord Translate [x y drawable]
    IOrigin
    (-origin [this]
        [x y])

  IBounds
  (-bounds [this]
      (let [[w h] (bounds drawable)
            [ox oy] (origin drawable)]
        [(+ w ox)
         (+ h oy)])))
```
The `-bounds` impl looks like it adds the origin `ox` to the bounds `w` and returns that as the new bounds. It looks like this code will return something with a larger bounds than the child drawable if the origin is positive. It looks like it changes the bounds of the child. Why? It doesn't seem like the bounds should change with a translate. Also, I can't actually observe this. i.e. The bounds do not change after a translate...
```clojure
(ui/bounds (ui/label "hi"))
;; [10.896484375 17.609375]
(ui/bounds (ui/translate 10 10 (ui/label "hi")))
;; [10.896484375 17.609375]
```
https://github.com/phronmophobic/membrane/blob/f00e3cd11a0c26a9650fd92111efcb9bb2ec2358/src/membrane/ui.cljc#L629

**Comment by phronmophobic** [comment]
> The bounds do not change after a translate...
The reason the bounds aren't affected is that it's adding the origin of the child element and not the origin of the translate element. In this example the label's origin is [0 0].

However, I think including `drawable`'s origin as part of its bounds might be a bug. Early on, the `bounds` of an element depended on the `origin`, but I think I transitioned to separating `bounds` and `origin` inconsistently. I should revisit this.

**Reply by phronmophobic** [reply]
I think the bug hasn't manifested very often since most elements have their origin at [0 0].

**Reply by rgkirch** [reply]
Ah. Huh. Thanks!

**Reply by phronmophobic** [reply]
I'm most certainly using this answers feature wrong.

Anyway, I've looked into this a bit more. The policy that I came up with for handling origin and bounds is the following:

1. `-bounds` should not include the current element's `-origin` offset in its calculation
2. When calculating the bounds of child elements, elements _should_ include the `-origin` offset.

I think 1) is somewhat intuitive, but 2) could use an explanation. 

The simplest element with child elements to analyze is a group element (ie. a vector). Suppose you have a group like the following:
```clojure
[(ui/rectangle 5 5)
 (ui/translate 5 5
     (ui/rectangle 5 5)]
```
What's the origin of this group element? Obviously, it's `[0 0]`. Ok, what are the bounds of the group element? I think the most reasonable answer is `[10 10]`. That means we're implicitly including the origin when calculating the bounds of the group's children and the group element's bounds are the bounds that encompass all of its children.

Now let's try a trickier example:
```clojure
[(ui/translate 5 5
     (ui/rectangle 5 5)]
```
I think it makes sense for groups to always have their origin at `[0 0]`. Given that, I think the bounds for this element should still be `[10 10]`.

Given these examples, you can apply the same logic to an element like `membrane.ui/padding` which only has one child element.

**Reply by phronmophobic** [reply]
See https://github.com/phronmophobic/membrane/commit/6631f71754ceca2b8abd35aca7b3a547f1fcef6f
